test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java

```diff
@@ -0,0 +1,93 @@
+package software.amazon.awssdk.services.waiters;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.core.retry.backoff.BackoffStrategy;
+import software.amazon.awssdk.core.waiters.PollingStrategy;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.services.restjsonwithwaiters.RestJsonWithWaitersClient;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesRequest;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesResponse;
+import software.amazon.awssdk.services.restjsonwithwaiters.waiters.RestJsonWithWaitersWaiter;
+
+public class WaitersSyncFunctionalTest {
```

Comment:
Seems like we need to test other scenarios as well such as 

 - What happens when we reach max number of attempts
 - What happens when we get an unexpected exception
 - What happens to the client and scheduled executor when the the waiter is closed?
 - etc.

cc: @zoewangg for more suggestions

Reply:
The max number of attempts case is added in sync Waiters. Unexpected exception case has been covered by the tests for generic Waiters. For the third case, if the client and scheduled executor are configured while creating the Waiters then they won't be closed while the Waiters are closed, and tests related to this case are also added.

Reply:
@dagnir Good call! I added the test cases with unexpected exception and caught a bug in waiterRuntime 😅

Reply:
😅Guess that's my fault. I misunderstood the unmatched exception case in the `BaseWaiterTest`, thanks for making sure about it again.